iree/compiler/Conversion/LinalgToLLVM/FastExpConversion.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+// Fast polynomial approximation of exp(x) using its reduced range exp(y)
+// where y is in the range [0, ln(2)], let y = x - floor(x / ln(2)) * ln(2)
+// = x - k * ln(2), exp(x) = exp(y) * 2^k. exp(y) is computed with 4th degree
+// polyomial: exp(y) = c0 + c1 * y + c2 * y^2 + c3 * y^3 + c4 * y^4
+struct FastExpConversionPattern : public OpRewritePattern<LLVM::ExpOp> {
+  using OpRewritePattern<LLVM::ExpOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(LLVM::ExpOp op,
+                                PatternRewriter &rewriter) const override {
+    constexpr float ln2Const = 0.69314718055994529;
+    constexpr float ln2InvConst = 1.4426950408889634;
+    constexpr float cVaues[5] = {0.05924867, 0.15514645, 0.50308552, 0.99968939,
+                                 1.0000072153251447};
```

Comment:
Can you explain in a comment how these cValues are computed?

(Side note: i checked if current compilers generate good code for `static const float k = std::log(2.0f)`, that is, not only evaluating k at compile time but not even having a lock around it. GCC does... but Clang does not! So you are doing it right (sadly) to use literals here). https://godbolt.org/z/dhd3Yb )

Reply:
They are least square fit computed with numpy I added a comment above